features/shard: delay unlink of a file that has fd_count > 0

vh05 opened the pull request:
When there are multiple processes working on a file and if any
process unlinks that file then unlink operation shouldn't harm
other processes working on it. This is a posix a compliant
behavior and this should be supported when shard feature is
enabled also.

Problem description:
Let's consider 2 clients C1 and C2 working on a file F1 with 5
shards on gluster mount and gluster server has 4 bricks
B1, B2, B3, B4.

Assume that base file/shard is present on B1, 1st, 2nd shards
on B2, 3rd and 4th shards on B3 and 5th shard falls on B4 C1
has opened the F1 in append mode and is writing to it. The
write FOP goes to 5th shard in this case. So the
inode->fd_count = 1 on B1(base file) and B4 (5th shard).

C2 at the same time issued unlink to F1. On the server, the
base file has fd_count = 1 (since C1 has opened the file),
the base file is renamed under .glusterfs/unlink and
returned to C2. Then unlink will be sent to shards on all
bricks and shards on B2 and B3 will be deleted which have
no open reference yet. C1 starts getting errors while
accessing the remaining shards though it has open references
for the file.

This is one such undefined behavior. Likewise we will
encounter many such undefined behaviors as we dont have one
global lock to access all shards as one. Of Course having such
global lock will lead to performance hit as it reduces window
for parallel access of shards.

Solution:
The above undefined behavior can be addressed by delaying the
unlink of a file when there are open references on it.
File unlink happens in 2 steps.
step 1: client creates marker file under .shard/remove_me and
sends unlink on base file to the server
step 2: on return from the server, the associated shards will
be cleaned up and finally marker file will be removed.

In step 2, the back ground deletion process does nameless
lookup using marker file name (marker file is named after the
gfid of the base file) in glusterfs/unlink dir. If the nameless
look up is successful then that means the gfid still has open
fds and deletion of shards has to be delayed. If nameless
lookup fails then that indicates the gfid is unlinked and no
open fds on that file (the gfid path is unlinked during final
close on the file). The shards on which deletion is delayed
are unlinked one the all open fds are closed and this is
done through a thread which wakes up every 10 mins.

Also removed active_fd_count from inode structure and
referring fd_count wherever active_fd_count was used.

fixes: #1358
Change-Id: I8985093386e26215e0b0dce294c534a66f6ca11c
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

vh05 committed:
features/shard: delay unlink of a file that has fd_count > 0

When there are multiple processes working on a file and if any
process unlinks that file then unlink operation shouldn't harm
other processes working on it. This is a posix a compliant
behavior and this should be supported when shard feature is
enabled also.

Problem description:
Let's consider 2 clients C1 and C2 working on a file F1 with 5
shards on gluster mount and gluster server has 4 bricks
B1, B2, B3, B4.

Assume that base file/shard is present on B1, 1st, 2nd shards
on B2, 3rd and 4th shards on B3 and 5th shard falls on B4 C1
has opened the F1 in append mode and is writing to it. The
write FOP goes to 5th shard in this case. So the
inode->fd_count = 1 on B1(base file) and B4 (5th shard).

C2 at the same time issued unlink to F1. On the server, the
base file has fd_count = 1 (since C1 has opened the file),
the base file is renamed under .glusterfs/unlink and
returned to C2. Then unlink will be sent to shards on all
bricks and shards on B2 and B3 will be deleted which have
no open reference yet. C1 starts getting errors while
accessing the remaining shards though it has open references
for the file.

This is one such undefined behavior. Likewise we will
encounter many such undefined behaviors as we dont have one
global lock to access all shards as one. Of Course having such
global lock will lead to performance hit as it reduces window
for parallel access of shards.

Solution:
The above undefined behavior can be addressed by delaying the
unlink of a file when there are open references on it.
File unlink happens in 2 steps.
step 1: client creates marker file under .shard/remove_me and
sends unlink on base file to the server
step 2: on return from the server, the associated shards will
be cleaned up and finally marker file will be removed.

In step 2, the back ground deletion process does nameless
lookup using marker file name (marker file is named after the
gfid of the base file) in glusterfs/unlink dir. If the nameless
look up is successful then that means the gfid still has open
fds and deletion of shards has to be delayed. If nameless
lookup fails then that indicates the gfid is unlinked and no
open fds on that file (the gfid path is unlinked during final
close on the file). The shards on which deletion is delayed
are unlinked one the all open fds are closed and this is
done through a thread which wakes up every 10 mins.

Also removed active_fd_count from inode structure and
referring fd_count wherever active_fd_count was used.

fixes: #1358
Change-Id: I8985093386e26215e0b0dce294c534a66f6ca11c
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

VHariharmath-rh commented:
/run regression

vh05 commented:
/run regression

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -7084,6 +7133,120 @@ shard_seek(call_frame_t *frame, xlator_t *this, fd_t *fd, off_t offset,
+        ret = pthread_join(ti->thread, NULL);
+        if (ret)
+            gf_msg(this->name, GF_LOG_WARNING, 0, 0,
+                   "Failed to clean up shard unlink thread.");
```
It's not important, but `ti->running` should be set to `_gf_false`, just for consistency.

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -7084,6 +7133,120 @@ shard_seek(call_frame_t *frame, xlator_t *this, fd_t *fd, off_t offset,
+        goto out;
+
+    ret = pthread_cond_init(&ti->cond, NULL);
+    if (ret)
```
ti->mutex should be destroyed in this case.

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -7084,6 +7133,120 @@ shard_seek(call_frame_t *frame, xlator_t *this, fd_t *fd, off_t offset,
+    }
+unlock:
+    pthread_mutex_unlock(&ti->mutex);
+    return -ret;
```
ret is already negative if there's some error creating the thread.

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -7084,6 +7133,120 @@ shard_seek(call_frame_t *frame, xlator_t *this, fd_t *fd, off_t offset,
+        } else {
+            ret = gf_thread_create(&ti->thread, NULL, shard_unlink_handler, ti,
+                                   "shard_unlink");
+            if (ret)
```
`ret < 0` would be better.

Also, if thread creation fails, the volume will continue working but shards won't be deleted. We should, at least, leave a log error so that this issue can be identified.

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -7150,6 +7313,13 @@ init(xlator_t *this)
+
+    ret = shard_unlink_handler_init(&priv->thread_info);
+    if (ret) {
+        ret = 0;
```
Why ? shouldn't we fail if it cannot be initialized ?

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -215,6 +220,28 @@ posix_lookup(call_frame_t *frame, xlator_t *this, loc_t *loc, dict_t *xdata)
+        /* The gfid will be renamed to ".glusterfs/unlink" in case
+         * there are any open fds on the file in posix_unlink path.
+         * So client can request server to do nameless lookup with
+         * xdata = GF_NAMELESS_UNLINK_DIR_LOOKUP in ".glusterfs/unlink"
```
```suggestion
         * xdata = GF_UNLINKED_LOOKUP in ".glusterfs/unlink"
```

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -215,6 +220,28 @@ posix_lookup(call_frame_t *frame, xlator_t *this, loc_t *loc, dict_t *xdata)
+        ret = dict_get_uint32(xdata, GF_UNLINKED_LOOKUP, &lookup_unlink_dir);
+        if (!ret && lookup_unlink_dir) {
+            POSIX_GET_FILE_UNLINK_PATH(priv->base_path, loc->gfid, unlink_path);
+            ret = sys_lstat(unlink_path, &lstatbuf);
```
Probably it would be better to use `buf` instead of `lstatbuf` so that we return a valid iatt if the file still exists. It could be useful for some cases. In this case we should set `ia_nlink = 0` to be consistent.

VHariharmath-rh replied:
My bad. Changing lstatbuf to buf. Btw, why "ia_nlink = 0" should be done when the file exists?

vh05 replied:
Please correct me if wrong. sys_lstat() wants struct stat which lstatbuf as 2nd arg and buf is of struct attr ?. What should be replaced here? The only intention here is to verify the file presence under unlink dir and should we construct iattr if file present under unlink dir ?

Also, in what case ia_nlink = 0 should be done ?

xhernandez replied:
If a client wants to lookup a file, it might be to use it. Since the file still exists, I think we should return all the information we can, at least the iatt. It's true that `buf` has a different type. We should convert `lstatbuf` to `struct iatt` in `buf` to return it.

I'm also not sure if it wouldn't be better to also return xattrs as regular lookups do, but for now I think iatt data is enough.

`ia_nlink` should always be set to 0, as the file is deleted so it doesn't have any link.

vh05 replied:
Okay. will do iatt_from_stat() and set ia_nlink to 0 in case of ret == ENOENT

xhernandez commented on xlators/storage/posix/src/posix-inode-fd-ops.c:
```diff
@@ -2529,6 +2570,9 @@ posix_release(xlator_t *this, fd_t *fd)
     VALIDATE_OR_GOTO(this, out);
     VALIDATE_OR_GOTO(fd, out);
 
+    if (fd->inode->fd_count == 1)
+        posix_unlink_renamed_file(this, fd->inode);
```
If the file has no open fd's and there's no entry pointing to the inode, the inode itself should be forgotten and the file unlinked in `posix_forget()`. Otherwise it means that we are holding an unnecessary reference and wasting memory (probably that's the case right now, so this is needed for now, but I would like you to add a comment here about this).

vh05 replied:
Thinking to rename "posix_unlink_renamed_file" to something like "posix_check_free_ctx" as prior also does the freeing of the context

xhernandez commented on xlators/storage/posix/src/posix-inode-fd-ops.c:
```diff
@@ -2529,6 +2570,9 @@ posix_release(xlator_t *this, fd_t *fd)
     VALIDATE_OR_GOTO(this, out);
     VALIDATE_OR_GOTO(fd, out);
 
+    if (fd->inode->fd_count == 1)
```
It's unfortunate that `release` function is called before decrementing the `fd_count` of the inode. This makes impossible to do this in a safe way. For example, if there are two concurrent releases, `fd_count` will be 2, so none of the threads will unlink the renamed file.

I think we should modify `fd_destroy()` in `libglusterfs/src/fd.c` so that the `fd_count` is updated before calling `release`, but there's a comment there to only decrease the count after having closed the file. I'm not sure if this means after having called `release`.

@pranithk you added that comment. Is it safe to decrement before calling `release` ? is there any known dependency ?

pranithk replied:
@xhernandez In inode there are 2 variables that keep track of fd-count.
This is my understanding
1) fd_count: Tracks the opens fds on the disk file system
     Is incremented on fd_bind() and decremented after fd_destroy()
2) active_fd_count: Tracks the fd_t structures that are active. This number will be decremented before calling fd_destroy(). I think eager-lock required this subtle difference.

pranithk replied:
@xhernandez I went through the code again just now. It doesn't look like we need these two with subtle differences. We can just keep the behavior of active-fd-count and things should work as expected. For backward compatibility we should send the same count in both GET_OPEN_FD_COUNT and GET_ACTIVE_OPEN_FD_COUNT

vh05 replied:
I have moved fd decrement part before calling release in fd_destroy() and ran shard suit. The suit was successful. I will push the changes and run the regression.

pranithk replied:
Could you also remove active_fd_count variable in inode and just keep fd_count?

vh05 replied:
sure @pranithk

vh05 replied:
@pranithk @xhernandez Changes are done to the fd_unref() and fd_destroy() function.  

Sorry that I forcefully updated the patch.

vh05 commented:
/run regression

vh05 commented:
/run regression

xhernandez commented on tests/bugs/shard/issue-1358.t:
```diff
@@ -0,0 +1,88 @@
+exec 5>&-
+
+# Check ".shard/.remove_me"
+EXPECT_WITHIN $FILE_COUNT_TIME 1 get_file_count $B0/${V0}0/.shard/.remove_me/$gfid_new
```
Testing this after having closed the file couldn't cause spurious failures if, for some reason, the thread that removes the entries has been running before this test ?

Wouldn't it be safer to test this before closing the file ?

vh05 replied:
Yes. That's right. I will make this change.

xhernandez commented on tests/bugs/shard/issue-1425.t:
```diff
@@ -21,7 +21,13 @@ TEST $CLI volume profile $V0 start
 
 TEST $GFS --volfile-id=$V0 --volfile-server=$H0 $M0
 
+echo `$CLI volume profile $V0 info incremental | grep -w LOOKUP | awk '{print $8}'` > /dev/zero
```
Is this really needed ? it doesn't seem to be used.

vh05 replied:
Yes. Just want to clear profile info to get the only profile numbers after fallocate which is done in next step.

I will replace this with "$CLI volume profile $V0 info clear" which does the job for us

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -3642,6 +3644,45 @@ shard_lookup_marker_entry(xlator_t *this, shard_local_t *local,
+        dict_unref(xattr_req);
+    loc_wipe(&loc);
+
+    return ret;
```
`ret` has different meanings. It can contain an error code or be simply -1 in case of error. I prefer to use error codes, but given that we depend on functions that only return -1 and we can't change them now, I think we should always return -1 in case of error.

vh05 replied:
okay. Makes sense. I will return -1 instead of specific error codes.

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -215,6 +220,28 @@ posix_lookup(call_frame_t *frame, xlator_t *this, loc_t *loc, dict_t *xdata)
+            }
+            goto nameless_lookup_unlink_dir_out;
+        }
+        /* If you are here means the call is for regular nameless
+         * lookup of a file */
```
I'm not sure if this order is the right one. Checking first unlink path means that most of the lookups will require an additional system call to check for unlinks, which is not a common case. I think we should first try the regular approach, and if it fails with ENOENT, then check the unlink directory.

vh05 replied:
I changed the order of these calls and it resulting in some undefined behavior

xhernandez replied:
Why ? we should call `MAKE_INODE_HANDLE()` and check if it failed (`op_ret == -1`). If so, and `errno` is `ENOENT` , then we should check if the lookup request wants to check for unlinked inodes.

xhernandez commented on xlators/storage/posix/src/posix-inode-fd-ops.c:
```diff
@@ -2519,6 +2519,47 @@ posix_flush(call_frame_t *frame, xlator_t *this, fd_t *fd, dict_t *xdata)
+    struct posix_private *priv = this->private;
+
+    ret = inode_ctx_del2(inode, this, &ctx_uint1, &ctx_uint2);
+    if (!ctx_uint1)
+        goto check_ctx2;
```
I think we can minimize the use of `goto` by rearranging the code a bit different. For example we can check here for `ctx_uint2` and free it if necessary and then continue with `ctx_uint1`, for example.

```c
    GF_FREE((posix_mdata_t *)(uintptr_t)ctx_uint2);

    if (ctx_uint1 == 0) {
        return 0;
    }

    ctx = (posix_inode_ctx_t *)(uintptr_t)ctx_uint1;
    /* ... */
```

vh05 replied:
sure

xhernandez commented on xlators/storage/posix/src/posix-inode-fd-ops.c:
```diff
@@ -2519,6 +2519,47 @@ posix_flush(call_frame_t *frame, xlator_t *this, fd_t *fd, dict_t *xdata)
+            gf_msg(this->name, GF_LOG_ERROR, ENOMEM, P_MSG_UNLINK_FAILED,
+                   "Failed to remove gfid :%s", uuid_utoa(inode->gfid));
+            ret = -1;
+            goto ctx_free;
+        }
+        ret = sys_unlink(unlink_path);
```
```suggestion
        } else {
            ret = sys_unlink(unlink_path);
        }
```

vh05 replied:
sure

xhernandez commented on xlators/storage/posix/src/posix-inode-fd-ops.c:
```diff
@@ -2529,6 +2570,9 @@ posix_release(xlator_t *this, fd_t *fd)
     VALIDATE_OR_GOTO(this, out);
     VALIDATE_OR_GOTO(fd, out);
 
+    if (fd->inode && fd->inode->fd_count == 0)
```
Can we have an 'fd' without an 'inode' ?

vh05 replied:
My bad. Just added that not having the thought behind it

xhernandez commented on xlators/storage/posix/src/posix-inode-fd-ops.c:
```diff
@@ -2529,6 +2570,9 @@ posix_release(xlator_t *this, fd_t *fd)
     VALIDATE_OR_GOTO(fd, out);
 
+    if (fd->inode && fd->inode->fd_count == 0)
+        posix_unlink_renamed_file_and_free_ctx(this, fd->inode);
```
Is it really necessary to do this here ? once the file is closed and the reference to the inode removed, shouldn't `posix_forget()` take care of the inode cleanup ?

vh05 replied:
Once the final reference to the file is closed/released, we receive a call to fd_destroy() ==> posix_release().
I have not observed posix_forget() being called when the file is closed. In case of unlink, we see posix_forget() is invoked and not seen posix_release() during unlink.

 I have captured call stack for both file close() and unlink

1. call stack for file close() ===>posix_release() is triggered

Thread 17 "glfs_rpcrqhnd" hit Breakpoint 1, posix_release (this=0x7fb4c000a140, fd=0x7fb49001aeb8) at posix-inode-fd-ops.c:2566
2566	    uint64_t tmp_pfd = 0;
(gdb) bt
#0  posix_release (this=0x7fb4c000a140, fd=0x7fb49001aeb8) at posix-inode-fd-ops.c:2566
#1  0x00007fb4d9a3ca05 in fd_destroy (fd=0x7fb49001aeb8) at fd.c:479
#2  fd_unref (fd=0x7fb49001aeb8) at fd.c:546
#3  0x00007fb4c702343b in server4_0_release (req=0x7fb4b8003488) at server-rpc-fops_v2.c:4042
#4  0x00007fb4d99ba025 in rpcsvc_request_handler (arg=0x7fb4c6f84140) at rpcsvc.c:2232
#5  0x00007fb4d97a4432 in start_thread () from /lib64/libpthread.so.0
#6  0x00007fb4d93e39d3 in clone () from /lib64/libc.so.6

2. call stack for unlink of a file() ===> posiz_forget() gets triggered
Thread 23 "glfs_iotwr006" hit Breakpoint 2, posix_forget (this=0x7fb4c000a140, inode=0x7fb4900144f8) at posix-inode-fd-ops.c:6005
6005	    return posix_unlink_renamed_file_and_free_ctx(this, inode);
(gdb) bt
#0  posix_forget (this=0x7fb4c000a140, inode=0x7fb4900144f8) at posix-inode-fd-ops.c:6005
#1  0x00007fb4d9a26b50 in __inode_ctx_free (inode=<optimized out>) at inode.c:348
#2  __inode_ctx_free (inode=0x7fb4900144f8) at inode.c:331
#3  0x00007fb4d9a28059 in __inode_destroy (inode=0x7fb4900144f8) at inode.c:1628
#4  inode_table_prune (table=table@entry=0x7fb4c007da50) at inode.c:1628
#5  0x00007fb4d9a281b8 in inode_unref (inode=0x7fb4900144f8) at inode.c:603
#6  0x00007fb4d9a15472 in loc_wipe (loc=loc@entry=0x7fb4b0027040) at xlator.c:1020
#7  0x00007fb4d9a8ea32 in args_wipe (args=0x7fb4b0027040) at default-args.c:1622
#8  0x00007fb4d9a3ba49 in call_stub_wipe_args (stub=0x7fb4b0026ff8) at call-stub.c:2367
#9  call_stub_destroy (stub=0x7fb4b0026ff8) at call-stub.c:2367
#10 0x00007fb4c7177740 in iot_worker (data=0x7fb4c0056570) at io-threads.c:232
#11 0x00007fb4d97a4432 in start_thread () from /lib64/libpthread.so.0
#12 0x00007fb4d93e39d3 in clone () from /lib64/libc.so.6

xhernandez replied:
The `fd` keeps a reference of the inode, so `posix_forget` cannot be called until `fd` has been completely destroyed. Even then, it might happen that other xlators keep a reference to the inode, so `posix_forget` may be called a bit later or even from another thread.

You should check if after continuing from `posix_release`, `posix_forget` is not called. If it's not called after some time, we might have an inode leak.

vh05 replied:
Yeah, It works as expected. Removing the unnecessary code from posix_release()

vh05 commented:
/run regression

vh05 commented:
/run regression

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -215,6 +220,32 @@ posix_lookup(call_frame_t *frame, xlator_t *this, loc_t *loc, dict_t *xdata)
+            if (ret) {
+                op_ret = -1;
+                op_errno = ret;
+                if (ret == ENOENT)
+                    buf.ia_nlink = 0;
```
In case of error `buf` is undefined, so we don't need to update it.

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -215,6 +220,32 @@ posix_lookup(call_frame_t *frame, xlator_t *this, loc_t *loc, dict_t *xdata)
+                if (ret == ENOENT)
+                    buf.ia_nlink = 0;
+            } else {
+                iatt_from_stat(&buf, &lstatbuf);
```
We need to set `buf.ia_nlink` to 0 here. The unlinked file exists, but there isn't any name referencing it, so the number of hardlinks should be set to 0.

xhernandez reviewed (changes requested):
When applying changes from comments, please don't amend the latest commit, create a new commit with only the changes. This makes review much much easier.

vh05 committed:
features/shard: delay unlink of a file that has fd_count > 0

When there are multiple processes working on a file and if any
process unlinks that file then unlink operation shouldn't harm
other processes working on it. This is a posix a compliant
behavior and this should be supported when shard feature is
enabled also.

Problem description:
Let's consider 2 clients C1 and C2 working on a file F1 with 5
shards on gluster mount and gluster server has 4 bricks
B1, B2, B3, B4.

Assume that base file/shard is present on B1, 1st, 2nd shards
on B2, 3rd and 4th shards on B3 and 5th shard falls on B4 C1
has opened the F1 in append mode and is writing to it. The
write FOP goes to 5th shard in this case. So the
inode->fd_count = 1 on B1(base file) and B4 (5th shard).

C2 at the same time issued unlink to F1. On the server, the
base file has fd_count = 1 (since C1 has opened the file),
the base file is renamed under .glusterfs/unlink and
returned to C2. Then unlink will be sent to shards on all
bricks and shards on B2 and B3 will be deleted which have
no open reference yet. C1 starts getting errors while
accessing the remaining shards though it has open references
for the file.

This is one such undefined behavior. Likewise we will
encounter many such undefined behaviors as we dont have one
global lock to access all shards as one. Of Course having such
global lock will lead to performance hit as it reduces window
for parallel access of shards.

Solution:
The above undefined behavior can be addressed by delaying the
unlink of a file when there are open references on it.
File unlink happens in 2 steps.
step 1: client creates marker file under .shard/remove_me and
sends unlink on base file to the server
step 2: on return from the server, the associated shards will
be cleaned up and finally marker file will be removed.

In step 2, the back ground deletion process does nameless
lookup using marker file name (marker file is named after the
gfid of the base file) in glusterfs/unlink dir. If the nameless
look up is successful then that means the gfid still has open
fds and deletion of shards has to be delayed. If nameless
lookup fails then that indicates the gfid is unlinked and no
open fds on that file (the gfid path is unlinked during final
close on the file). The shards on which deletion is delayed
are unlinked one the all open fds are closed and this is
done through a thread which wakes up every 10 mins.

Also removed active_fd_count from inode structure and
referring fd_count wherever active_fd_count was used.

fixes: #1358

Change-Id: Iec16d7ff5e05f29255491a43fbb6270c72868999
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

vh05 commented:
/run regression

vh05 committed:
features/shard: delay unlink of a file that has fd_count > 0

When there are multiple processes working on a file and if any
process unlinks that file then unlink operation shouldn't harm
other processes working on it. This is a posix a compliant
behavior and this should be supported when shard feature is
enabled also.

Problem description:
Let's consider 2 clients C1 and C2 working on a file F1 with 5
shards on gluster mount and gluster server has 4 bricks
B1, B2, B3, B4.

Assume that base file/shard is present on B1, 1st, 2nd shards
on B2, 3rd and 4th shards on B3 and 5th shard falls on B4 C1
has opened the F1 in append mode and is writing to it. The
write FOP goes to 5th shard in this case. So the
inode->fd_count = 1 on B1(base file) and B4 (5th shard).

C2 at the same time issued unlink to F1. On the server, the
base file has fd_count = 1 (since C1 has opened the file),
the base file is renamed under .glusterfs/unlink and
returned to C2. Then unlink will be sent to shards on all
bricks and shards on B2 and B3 will be deleted which have
no open reference yet. C1 starts getting errors while
accessing the remaining shards though it has open references
for the file.

This is one such undefined behavior. Likewise we will
encounter many such undefined behaviors as we dont have one
global lock to access all shards as one. Of Course having such
global lock will lead to performance hit as it reduces window
for parallel access of shards.

Solution:
The above undefined behavior can be addressed by delaying the
unlink of a file when there are open references on it.
File unlink happens in 2 steps.
step 1: client creates marker file under .shard/remove_me and
sends unlink on base file to the server
step 2: on return from the server, the associated shards will
be cleaned up and finally marker file will be removed.

In step 2, the back ground deletion process does nameless
lookup using marker file name (marker file is named after the
gfid of the base file) in glusterfs/unlink dir. If the nameless
look up is successful then that means the gfid still has open
fds and deletion of shards has to be delayed. If nameless
lookup fails then that indicates the gfid is unlinked and no
open fds on that file (the gfid path is unlinked during final
close on the file). The shards on which deletion is delayed
are unlinked one the all open fds are closed and this is
done through a thread which wakes up every 10 mins.

Also removed active_fd_count from inode structure and
referring fd_count wherever active_fd_count was used.

fixes: #1358

Change-Id: I07e5a5bf9d33c24b63da72d4f3f59392c5421652
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

vh05 commented:
/run regression

vh05 commented:
/run regression

vh05 commented:
/run regression

vh05 committed:
features/shard: delay unlink of a file that has fd_count > 0

When there are multiple processes working on a file and if any
process unlinks that file then unlink operation shouldn't harm
other processes working on it. This is a posix a compliant
behavior and this should be supported when shard feature is
enabled also.

Problem description:
Let's consider 2 clients C1 and C2 working on a file F1 with 5
shards on gluster mount and gluster server has 4 bricks
B1, B2, B3, B4.

Assume that base file/shard is present on B1, 1st, 2nd shards
on B2, 3rd and 4th shards on B3 and 5th shard falls on B4 C1
has opened the F1 in append mode and is writing to it. The
write FOP goes to 5th shard in this case. So the
inode->fd_count = 1 on B1(base file) and B4 (5th shard).

C2 at the same time issued unlink to F1. On the server, the
base file has fd_count = 1 (since C1 has opened the file),
the base file is renamed under .glusterfs/unlink and
returned to C2. Then unlink will be sent to shards on all
bricks and shards on B2 and B3 will be deleted which have
no open reference yet. C1 starts getting errors while
accessing the remaining shards though it has open references
for the file.

This is one such undefined behavior. Likewise we will
encounter many such undefined behaviors as we dont have one
global lock to access all shards as one. Of Course having such
global lock will lead to performance hit as it reduces window
for parallel access of shards.

Solution:
The above undefined behavior can be addressed by delaying the
unlink of a file when there are open references on it.
File unlink happens in 2 steps.
step 1: client creates marker file under .shard/remove_me and
sends unlink on base file to the server
step 2: on return from the server, the associated shards will
be cleaned up and finally marker file will be removed.

In step 2, the back ground deletion process does nameless
lookup using marker file name (marker file is named after the
gfid of the base file) in glusterfs/unlink dir. If the nameless
look up is successful then that means the gfid still has open
fds and deletion of shards has to be delayed. If nameless
lookup fails then that indicates the gfid is unlinked and no
open fds on that file (the gfid path is unlinked during final
close on the file). The shards on which deletion is delayed
are unlinked one the all open fds are closed and this is
done through a thread which wakes up every 10 mins.

Also removed active_fd_count from inode structure and
referring fd_count wherever active_fd_count was used.

fixes: #1358

Change-Id: I3679de8545f2e5b8027c4d5a6fd0592092e8dfbd
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

vh05 commented:
/run regression

vh05 commented:
/run regression

xhernandez commented on xlators/storage/posix/src/posix-entry-ops.c:
```diff
@@ -230,23 +232,23 @@ posix_lookup(call_frame_t *frame, xlator_t *this, loc_t *loc, dict_t *xdata)
+                ret = sys_lstat(unlink_path, &lstatbuf);
+                if (ret) {
+                    op_ret = -1;
+                    op_errno = ret;
```
```suggestion
                    op_errno = errno;
```

vh05 commented:
/run regression

vh05 committed:
Update xlators/storage/posix/src/posix-entry-ops.c

Co-authored-by: Xavi Hernandez <xhernandez@users.noreply.github.com>

vh05 committed:
features/shard: delay unlink of a file that has fd_count > 0

When there are multiple processes working on a file and if any
process unlinks that file then unlink operation shouldn't harm
other processes working on it. This is a posix a compliant
behavior and this should be supported when shard feature is
enabled also.

Problem description:
Let's consider 2 clients C1 and C2 working on a file F1 with 5
shards on gluster mount and gluster server has 4 bricks
B1, B2, B3, B4.

Assume that base file/shard is present on B1, 1st, 2nd shards
on B2, 3rd and 4th shards on B3 and 5th shard falls on B4 C1
has opened the F1 in append mode and is writing to it. The
write FOP goes to 5th shard in this case. So the
inode->fd_count = 1 on B1(base file) and B4 (5th shard).

C2 at the same time issued unlink to F1. On the server, the
base file has fd_count = 1 (since C1 has opened the file),
the base file is renamed under .glusterfs/unlink and
returned to C2. Then unlink will be sent to shards on all
bricks and shards on B2 and B3 will be deleted which have
no open reference yet. C1 starts getting errors while
accessing the remaining shards though it has open references
for the file.

This is one such undefined behavior. Likewise we will
encounter many such undefined behaviors as we dont have one
global lock to access all shards as one. Of Course having such
global lock will lead to performance hit as it reduces window
for parallel access of shards.

Solution:
The above undefined behavior can be addressed by delaying the
unlink of a file when there are open references on it.
File unlink happens in 2 steps.
step 1: client creates marker file under .shard/remove_me and
sends unlink on base file to the server
step 2: on return from the server, the associated shards will
be cleaned up and finally marker file will be removed.

In step 2, the back ground deletion process does nameless
lookup using marker file name (marker file is named after the
gfid of the base file) in glusterfs/unlink dir. If the nameless
look up is successful then that means the gfid still has open
fds and deletion of shards has to be delayed. If nameless
lookup fails then that indicates the gfid is unlinked and no
open fds on that file (the gfid path is unlinked during final
close on the file). The shards on which deletion is delayed
are unlinked one the all open fds are closed and this is
done through a thread which wakes up every 10 mins.

Also removed active_fd_count from inode structure and
referring fd_count wherever active_fd_count was used.

fixes: #1358
Change-Id: I8985093386e26215e0b0dce294c534a66f6ca11c
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

vh05 commented:
/run regression

vh05 commented:
/recheck smoke

vh05 commented:
/run regression

vh05 commented:
/recheck smoke

vh05 commented:
/run regression

vh05 commented:
The regression test suite is failing on Jenkins run: https://build.gluster.org/job/gh_centos7-regression/691/console
Failed tests:
The below test is continuously failing.
./tests/00-geo-rep/georep-basic-tarssh-ec.t

These 2 tests are intermittently failing
./tests/bugs/glusterd/brick-mux.t
./tests/basic/afr/afr-no-fsync.t

But these tests are passing on my local setup. 

```
# prove -rfv tests/00-geo-rep/georep-basic-tarssh-ec.t
..
All tests are successful.
Files=1, Tests=81, 139 wallclock secs ( 0.05 usr  0.02 sys +  4.63 cusr  1.81 csys =  6.51 CPU)
Result: PASS

# prove -rfv ./tests/bugs/glusterd/brick-mux.t
..
ok  34 [   5038/    423] <  79> '2 online_brick_count'
ok
All tests successful.
Files=1, Tests=34, 35 wallclock secs ( 0.03 usr  0.01 sys +  4.25 cusr  1.37 csys =  5.66 CPU)
Result: PASS

# prove -rfv tests/basic/afr/afr-no-fsync.t
..
ok  10 [    696/    177] <  18> '! gluster --mode=script --wignore volume profile patchy info incremental | grep FSYNC'
ok
All tests successful.
Files=1, Tests=10, 18 wallclock secs ( 0.02 usr  0.00 sys +  0.59 cusr  0.67 csys =  1.28 CPU)
Result: PASS
```

None of the tests are related to my patch changes

xhernandez commented:
/run regression

xhernandez commented:
/run regression

xhernandez commented:
tests/features/trash.t seems to be crashing on unlink path, so it's probably related to this patch. Can you check it ?

vh05 commented:
/run regression

vh05 commented:
/recheck smoke

vh05 committed:
Update fd.c

VHariharmath-rh commented:
/run regression

xhernandez commented:
/run regression

vh05 committed:
features/shard: delay unlink of a file that has fd_count > 0

When there are multiple processes working on a file and if any
process unlinks that file then unlink operation shouldn't harm
other processes working on it. This is a posix a compliant
behavior and this should be supported when shard feature is
enabled also.

Problem description:
Let's consider 2 clients C1 and C2 working on a file F1 with 5
shards on gluster mount and gluster server has 4 bricks
B1, B2, B3, B4.

Assume that base file/shard is present on B1, 1st, 2nd shards
on B2, 3rd and 4th shards on B3 and 5th shard falls on B4 C1
has opened the F1 in append mode and is writing to it. The
write FOP goes to 5th shard in this case. So the
inode->fd_count = 1 on B1(base file) and B4 (5th shard).

C2 at the same time issued unlink to F1. On the server, the
base file has fd_count = 1 (since C1 has opened the file),
the base file is renamed under .glusterfs/unlink and
returned to C2. Then unlink will be sent to shards on all
bricks and shards on B2 and B3 will be deleted which have
no open reference yet. C1 starts getting errors while
accessing the remaining shards though it has open references
for the file.

This is one such undefined behavior. Likewise we will
encounter many such undefined behaviors as we dont have one
global lock to access all shards as one. Of Course having such
global lock will lead to performance hit as it reduces window
for parallel access of shards.

Solution:
The above undefined behavior can be addressed by delaying the
unlink of a file when there are open references on it.
File unlink happens in 2 steps.
step 1: client creates marker file under .shard/remove_me and
sends unlink on base file to the server
step 2: on return from the server, the associated shards will
be cleaned up and finally marker file will be removed.

In step 2, the back ground deletion process does nameless
lookup using marker file name (marker file is named after the
gfid of the base file) in glusterfs/unlink dir. If the nameless
look up is successful then that means the gfid still has open
fds and deletion of shards has to be delayed. If nameless
lookup fails then that indicates the gfid is unlinked and no
open fds on that file (the gfid path is unlinked during final
close on the file). The shards on which deletion is delayed
are unlinked one the all open fds are closed and this is
done through a thread which wakes up every 10 mins.

Also removed active_fd_count from inode structure and
referring fd_count wherever active_fd_count was used.

fixes: #1358
Change-Id: I8985093386e26215e0b0dce294c534a66f6ca11c
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

VHariharmath-rh commented:
/run regression

vh05 commented:
/recheck smoke

vh05 commented:
/run regression

vh05 commented:
/run regression

vh05 commented:
/run regression

vh05 commented:
/recheck smoke

vh05 commented:
/run regression

vh05 commented:
/run regression

vh05 commented:
/run regression

vh05 commented:
/run regression

vh05 commented:
/run regression

deepshikhaaa commented:
/run regression

deepshikhaaa commented:
/run regression

deepshikhaaa commented:
/run regression

deepshikhaaa commented:
/run regression

vh05 committed:
features/shard: delay unlink of a file that has fd_count > 0

When there are multiple processes working on a file and if any
process unlinks that file then unlink operation shouldn't harm
other processes working on it. This is a posix a compliant
behavior and this should be supported when shard feature is
enabled also.

Problem description:
Let's consider 2 clients C1 and C2 working on a file F1 with 5
shards on gluster mount and gluster server has 4 bricks
B1, B2, B3, B4.

Assume that base file/shard is present on B1, 1st, 2nd shards
on B2, 3rd and 4th shards on B3 and 5th shard falls on B4 C1
has opened the F1 in append mode and is writing to it. The
write FOP goes to 5th shard in this case. So the
inode->fd_count = 1 on B1(base file) and B4 (5th shard).

C2 at the same time issued unlink to F1. On the server, the
base file has fd_count = 1 (since C1 has opened the file),
the base file is renamed under .glusterfs/unlink and
returned to C2. Then unlink will be sent to shards on all
bricks and shards on B2 and B3 will be deleted which have
no open reference yet. C1 starts getting errors while
accessing the remaining shards though it has open references
for the file.

This is one such undefined behavior. Likewise we will
encounter many such undefined behaviors as we dont have one
global lock to access all shards as one. Of Course having such
global lock will lead to performance hit as it reduces window
for parallel access of shards.

Solution:
The above undefined behavior can be addressed by delaying the
unlink of a file when there are open references on it.
File unlink happens in 2 steps.
step 1: client creates marker file under .shard/remove_me and
sends unlink on base file to the server
step 2: on return from the server, the associated shards will
be cleaned up and finally marker file will be removed.

In step 2, the back ground deletion process does nameless
lookup using marker file name (marker file is named after the
gfid of the base file) in glusterfs/unlink dir. If the nameless
look up is successful then that means the gfid still has open
fds and deletion of shards has to be delayed. If nameless
lookup fails then that indicates the gfid is unlinked and no
open fds on that file (the gfid path is unlinked during final
close on the file). The shards on which deletion is delayed
are unlinked one the all open fds are closed and this is
done through a thread which wakes up every 10 mins.

Also removed active_fd_count from inode structure and
referring fd_count wherever active_fd_count was used.

fixes: #1358
Change-Id: I8985093386e26215e0b0dce294c534a66f6ca11c
Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

vh05 commented:
/run regression

xhernandez commented:
I'll wait till tomorrow before merging the patch to allow other to complete review.
